src/lora_region.erl

```diff
@@ -0,0 +1,1627 @@
+                FOptsOut;
+            ChMask ->
+                [{link_adr_req, datar_to_dr(Region, DataRate), TXPower, ChMask, Idx, 0} | FOptsOut]
+        end
+    ).
+
+%% transmission time estimation
+
+tx_time(FOpts, FRMPayloadSize, TxQ) ->
+    tx_time(phy_payload_size(FOpts, FRMPayloadSize), TxQ).
+
+phy_payload_size(FOpts, FRMPayloadSize) ->
+    1 + 7 + byte_size(FOpts) + 1 + FRMPayloadSize + 4.
+
+tx_time(PhyPayloadSize, #txq{datr = DataRate, codr = CodingRate}) ->
+    {SF, BW} = datar_to_tuple(DataRate),
+    {4, CR} = codr_to_tuple(CodingRate),
+    tx_time(PhyPayloadSize, SF, CR, BW * 1000).
+
+%% see http://www.semtech.com/images/datasheet/LoraDesignGuide_STD.pdf
+tx_time(PL, SF, CR, 125000) when SF == 11; SF == 12 ->
+    %% Optimization is mandated with spreading factors of 11 and 12 at 125 kHz bandwidth
+    tx_time(PL, SF, CR, 125000, 1);
+tx_time(PL, SF, CR, 250000) when SF == 12 ->
+    %% The firmware uses also optimization for SF 12 at 250 kHz bandwidth
+    tx_time(PL, SF, CR, 250000, 1);
+tx_time(PL, SF, CR, BW) ->
+    tx_time(PL, SF, CR, BW, 0).
+
+tx_time(PL, SF, CR, BW, DE) ->
+    TSym = math:pow(2, SF) / BW,
+    %% lorawan uses an explicit header
+    PayloadSymbNb = 8 + max(ceiling((8 * PL - 4 * SF + 28 + 16) / (4 * (SF - 2 * DE))) * CR, 0),
+    %% lorawan uses 8 symbols preamble
+    %% the last +1 is a correction based on practical experiments
+    1000 * ((8 + 4.25) + PayloadSymbNb + 1) * TSym.
+
+ceiling(X) ->
+    T = erlang:trunc(X),
+    case (X - T) of
+        Neg when Neg < 0 -> T;
+        Pos when Pos > 0 -> T + 1;
+        _ -> T
+    end.
+
+%% ------------------------------------------------------------------
+%% EUNIT Tests
+%% ------------------------------------------------------------------
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+
+channel_plan(ID) ->
+    case ID of
+        1 -> 'EU868';
+        2 -> 'US915';
+        3 -> 'CN779';
+        4 -> 'EU433';
+        5 -> 'AU915';
+        6 -> 'CN470';
+        7 -> 'AS923_1';
+        8 -> 'AS923_2';
+        9 -> 'AS923_3';
+        10 -> 'KR920';
+        11 -> 'IN865';
+        12 -> 'RU864';
+        13 -> 'AS923_4'
+    end.
+
+all_channel_plans_test() ->
+    [channel_plan(X) || X <- [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]].
+
+print_channel(Region, Ch) ->
+    Freq = uch2f(Region, Ch),
+    % io:format("Region= ~w Ch= ~w Freq= ~w~n", [Region, Ch, Freq]),
+    Freq.
+
+region_channels(Region) ->
+    [print_channel(Region, X) || X <- [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]].
+
+valid_channel(Region, Ch) ->
+    % io:format("Region = ~w Ch=~w ~n", [Region, Ch]),
+    #{min := Min, max := Max} = freq(Region),
+    Freq = uch2f(Region, Ch),
+    % io:format("Freq = ~w~n", [Freq]),
+    ?assert(Freq =< Max),
+    ?assert(Freq >= Min),
+    Ch2 = f2uch(Region, Freq),
+    % io:format("Ch2 = ~w~n", [Ch2]),
+    ?assertEqual(Ch, Ch2).
+
+valid_channel(Ch) ->
+    % valid_channel('EU868', Ch),
```

Comment:
Apologies, I should have mentioned.  There is no need to review lora_region.   Lora_region is only used internally for testing and it is an exact copy of the original Router Region code.

Reply:
Since `lora_region.erl` is used only for regression testing, please move it from `src/` to `test/`.

That will eliminate false positives when trying to discover name of library functions when using `grep` or `rg` (ripgrep).

Reply:
Following-up from Discord DM: 
Since there may be issues with stashing such a library under `test` subdirectory, maybe use a descriptive path like `src/regression-test-library_not-for-new-code/` instead, which should be obvious when using `grep` or `rg` (ripgrep).

Reply:
Better yet, use a combination of **both subdirectory and module name** indicating that it's intended only for regression testing.

Reply:
lora_region is now renamed test_region.
Also test_region is now protected with EUNIIT ifdef